eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala

```diff
@@ -0,0 +1,74 @@
+package fr.acinq.eclair.io
+
+import akka.actor.typed.delivery.DurableProducerQueue.TimestampMillis
+import akka.actor.typed.eventstream.EventStream
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.{ActorRef, Behavior}
+import fr.acinq.bitcoin.scalacompat.Crypto.PublicKey
+import fr.acinq.eclair.NodeParams
+import fr.acinq.eclair.channel.ChannelCreated
+import fr.acinq.eclair.io.Monitoring.Metrics
+import fr.acinq.eclair.io.Peer.Disconnect
+
+/**
+ * A singleton actor that limits the total number of incoming connections from peers that do not have channels with us.
+ *
+ * When a new incoming connection request is received, the Switchboard should send an
+ * [[IncomingConnectionsTracker.TrackIncomingConnection]] message.
+ *
+ * When the number of tracked peers exceeds `eclair.peer-connection.max-no-channels`, send [[Peer.Disconnect]] to
+ * the tracked peer with the oldest incoming connection.
+ *
+ * When a tracked peer disconnects or adds a channel, we will stop tracking that peer.
+ *
+ * We do not need to track peers that disconnect because they will terminate if they have no channels.
+ * Likewise, peers with channels will disconnect and terminate when their last channel closes.
+ *
+ * Note: Peers on the sync whitelist are not tracked.
+*/
+object IncomingConnectionsTracker {
+  // @formatter:off
+  sealed trait Command
+
+  case class TrackIncomingConnection(remoteNodeId: PublicKey) extends Command
+  private case class ForgetIncomingConnection(remoteNodeId: PublicKey) extends Command
+  private[io] case class CountIncomingConnections(replyTo: ActorRef[Int]) extends Command
+  // @formatter:on
+
+  def apply(nodeParams: NodeParams, switchboard: ActorRef[Disconnect]): Behavior[Command] = {
+    Behaviors.setup { context =>
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[PeerDisconnected](c => ForgetIncomingConnection(c.nodeId)))
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelCreated](c => ForgetIncomingConnection(c.remoteNodeId)))
```

Comment:
The issue is that `ChannelCreated` doesn't mean the channel opening flow will complete, a malicious peer will simply send `open_channel` and then do nothing, and that will let them be forgotten from this rate-limit.

Here is what I think we should do, let me know if you think this is a good idea. We remove this line, and only listen to `PeerDisconnected` events. Once we've reached our rate-limit, we choose a peer that we want to evict (with your current code, the one with the oldest connection) and:

- create a short-lived actor that will query its channel states (a bit like what the `PeerReadyNotifier` does)
- if any of that peer's channels has reached at least the `WAIT_FOR_FUNDING_CONFIRMED` state, which means that there is really a channel and we should keep the connection, we do a `ForgetIncomingConnection` for this peer and start trying to evict the next one in the list with the same process
- if that peer has no channel or none of his channels have made enough progress, we disconnect it

The benefit of that is that it complements nicely the work you've done on limiting the number of pending channels, because that creates a mechanism that will actively close connections with pending channels once we've reached our rate-limit. What do you think about that?

Reply:
I see what you mean. There's  currently no mechanism to terminate incoming channel open requests that do not proceed to the `WAIT_FOR_FUNDING_CONFIRMED` stage. But I believe this should be handled by the channel FSM, not a separate actor.

I feel like it would be easier to reason about this if `IncomingConnectionsTracker` did not need to query the internal state of a channel actor, but only know that there was a hand-off to a channel actor via  `ChannelCreated`. It would then be up to the channel actor to cancel the open request and kill itself if the initiator becomes non-responsive. When the last channel of a `Peer` terminates, the existing logic would kick in that terminates the `Peer` and closes its `PeerConnection`. Once a connection begins the process of creating a channel, it should succeed or close due to a timeout.

An attacker that opens a connection and then requests to open a channel could create free connections that would not count against the `peer-connection.max-no-channels` limit, but it would still be limited by `eclair.channel.maxTotalPendingChannelsPrivateNodes` because `PendingChannelsRateLimiter` limits the number of pending channels that have not completed. An attacker could only freely create connections up to the sum of these limits, not an unlimited number of connections.

Currently an ***initiator*** of channel open requests can set `OpenChannel.timeout_opt` to terminate the process if the counter-party becomes unresponsive - but only up until it reaches a committed state; either `WAIT_FOR_DUAL_FUNDING_CONFIRMED` or `WAIT_FOR_FUNDING_CONFIRMED`.

Why can't we also use this timeout mechanism as a ***non-initiator*** ? Especially in the case of dual-funding, the distinction between being an initiator or not is less clear when both sides may be adding funds.

Reply:
+1 to not increasing the complexity in the Channel FSM.

I like your idea of listening for `ChannelOpened`! It did not occur to me before that it's not a problem to disconnect a peer while they wait for on-chain confirmation and that the window of time while negotiating should normally be brief with honest peers.  Unless a node is being actively DoS'd (a limitation previously mentioned for this PR) a channel open should always be able to finish negotiation before being closed.

Fixed in [d625549](https://github.com/ACINQ/eclair/pull/2601/commits/d62554966e18aaf2521cb4403d25f87e41f8c292)

Reply:
It would be useful to make a couple of manual e2e tests to check how the `ReconnectionTask` interacts with that. I believe that it will make eclair automatically create an _outgoing_ connection (based on a random timer) if we evict a connection where we have a channel waiting for the funding transaction to confirm. That behavior would be a good thing.

Can you test the following scenarios and let me know how eclair behaves?

- Scenario 1
  - Accept an incoming connection which starts opening a channel
  - Simulate a malicious/buggy peer that stops contributing before we reach the state where we're waiting for the transaction to confirm (you'll have to modify eclair to avoid sending `funding_signed` on the fundee side for example)
  - Evict that connection (by having another node connect for example)
  - Verify that eclair does **not** automatically reconnect out to the malicious/buggy peer, even after the `ReconnectionTask` timer
- Scenario 2
  - Accept an incoming connection which starts opening a channel
  - Reach the state where we're waiting for the funding transaction to confirm (both peers are honest)
  - Evict that connection (by having another node connect for example)
  - Verify that eclair automatically reconnects **out** to that peer, which thus bypasses the rate limits (because it's an outgoing connection)